### PR TITLE
Add test for libs directory

### PR DESCRIPTION
**Pull request by HuseinA**

### CODE REVIEW

File: tests/test_admin.py [hunk 1]
```diff
@@ -58,6 +72,26 @@ def test_ls(monkeypatch, client):
         ],
     )
 
+    assert result.output == (
```
Comment (azzamsa) [inline tests/test_admin.py]: Put it in separate test function, with relevant function name corresponds the test goal.

File: tests/test_admin.py [hunk 2]
```diff
@@ -80,14 +114,18 @@ def fake_info(client, user_id, group_id):
...
 def test_info(monkeypatch, client):
-    monkeypatch.setattr(obs.libs.user, "info", fake_info)
-
     runner = CliRunner()
     result = runner.invoke(
-        cli,
-        ["admin", "user", "info", "--user-id", "StageTest", "--group-id", "testing"],
+        cli, ["admin", "user", "info", "--user-id", "StageTest", "--group-id", "tesng"]
+    )
+    assert result.output == (
```
Comment (azzamsa) [inline tests/test_admin.py]: idem

File: tests/test_admin.py [hunk 3]
```diff
@@ -116,13 +154,21 @@ def fake_qos_info(client, user_id, group_id):
...
 
 def test_qos_info(monkeypatch, client):
-    monkeypatch.setattr(obs.libs.qos, "info", fake_qos_info)
-
     runner = CliRunner()
     result = runner.invoke(
         cli, ["admin", "qos", "info", "--user-id", "StageTest", "--group-id", "testing"]
     )
 
+    assert result.output == (
```
Comment (azzamsa) [inline tests/test_admin.py]: idem

File: tests/test_admin.py [hunk 4]
```diff
@@ -144,10 +190,17 @@ def fake_human_date(unixtime):
 
 
 def test_cred_list(monkeypatch, client):
+    runner = CliRunner()
+    result = runner.invoke(
+        cli, ["admin", "cred", "ls", "--user-id", "StageTest", "--group-id", "testing"]
+    )
+    assert result.output == (
```
Comment (azzamsa) [inline tests/test_admin.py]: idem

File: tests/test_storage.py [hunk 5]
```diff
@@ -57,33 +61,51 @@ def fake_get_objects(resource, bucket_name, prefix=""):
...
     return [obj1, obj2]
 
 
 def test_ls_storage(monkeypatch, resource):
-    monkeypatch.setattr(obs.libs.bucket, "get_objects", fake_get_objects)
-
     runner = CliRunner()
     result = runner.invoke(cli, ["storage", "ls", "bucket-one"])
 
+    assert result.output == (f"'NoneType' object has no attribute 'Bucket'\n")
```
Comment (azzamsa) [inline tests/test_storage.py]: idem

File: tests/test_storage.py [hunk 6]
```diff
@@ -57,33 +61,51 @@ def fake_get_objects(resource, bucket_name, prefix=""):
...
 
 
 def test_bucket_usage(monkeypatch, resource):
-    monkeypatch.setattr(obs.libs.bucket, "get_objects", fake_get_objects)
-
     runner = CliRunner()
     result = runner.invoke(cli, ["storage", "du", "bucket-one"])
 
+    assert result.output == (
+        f"Bucket usage fetching failed. \n"
```
Comment (azzamsa) [inline tests/test_storage.py]: idem

File: tests/test_storage.py [hunk 7]
```diff
@@ -102,11 +124,14 @@ def fake_bucket_info(resource, bucket_name, auth):
 
 
 def test_bucket_info(monkeypatch, resource, plain_auth):
-    monkeypatch.setattr(obs.libs.bucket, "bucket_info", fake_bucket_info)
-
     runner = CliRunner()
```
Comment (azzamsa) [inline tests/test_storage.py]: idem

File: tests/test_storage.py [hunk 8]
```diff
@@ -134,11 +159,16 @@ def fake_object_info(resource, bucket_name, object_name):
 
 
 def test_object_info(monkeypatch, resource, plain_auth):
-    monkeypatch.setattr(obs.libs.bucket, "object_info", fake_object_info)
-
     runner = CliRunner()
     result = runner.invoke(cli, ["storage", "info", "bucket-one", "logo.png"])
 
+    assert result.output == (
+        f"Info fetching failed. \n" f"'NoneType' object has no attribute 'Object'\n"
```
Comment (azzamsa) [inline tests/test_storage.py]: idem

File: tests/test_storage.py [hunk 9]
```diff
@@ -154,12 +184,97 @@ def fake_presign(resource, bucket_name, object_name, expire):
 
 
 def test_presign(monkeypatch, resource):
-    monkeypatch.setattr(obs.libs.bucket, "generate_url", fake_presign)
-
     runner = CliRunner()
     result = runner.invoke(cli, ["storage", "presign", "bucket-one", "logo.png"])
 
+    assert (
+        result.output == f"URL generation failed. \n"
```
Comment (azzamsa) [inline tests/test_storage.py]: idem

File: tests/test_storage.py [hunk 10]
```diff
@@ -154,12 +184,97 @@ def fake_presign(resource, bucket_name, object_name, expire):
...
+        f'800.00 Byte, 3 objects in "black" bucket\n'
+        f"---\n"
+        f"1.07 KiB Total\n"
+    )
+
+
+def fake_gmt():
+    policies = {
+        "Jakarta": {"id": "123", "desc": "", "_": ""},
+        "Sydney": {"id": "143", "desc": "blabla", "_": ""},
```
Comment (azzamsa) [inline tests/test_storage.py]: foo/bar/foobar

**Comment by azzamsa** [comment]
Looks good to me.

**Comment by anak10thn** [comment]
still WIP @HuseinA ?

**Comment by HuseinA** [comment]
@anak10thn Done. Just changed it.

**Pull request merged**